### PR TITLE
SNOW-701257: optionally disable logging error for sql exceptions

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-701257
   Based on prior discussion here: https://github.com/snowflakedb/snowpark-python/pull/654, I am disabling logging on failed query be default and adding a parameter `log_on_exception` in `run_query` to log error message when needed.

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Comment by sfc-gh-sfan** [comment]
This should wait until we have the failed query info in the exception I think? Otherwise the information might be lost. Not sure if @sfc-gh-yixie has already made the change.

**Comment by sfc-gh-yixie** [comment]
@sfc-gh-sfan Yeah I talked to Afroz offline to add the failed query.

> This should wait until we have the failed query info in the exception I think? Otherwise the information might be lost. Not sure if @sfc-gh-yixie has already made the change.

File: tests/unit/test_server_connection.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def test_run_query_exceptions(mock_server_connection, caplog):
     ):
         with pytest.raises(Exception, match="fake exception"):
             with caplog.at_level(logging.ERROR):
-                mock_server_connection.run_query("fake query")
+                mock_server_connection.run_query("fake query", log_on_exception=True)
```
Comment (sfc-gh-jdu) [inline tests/unit/test_server_connection.py:89]: 1) Can we enable it for all tests we have?
2) How does a user enable log on exception? `collect(log_on_exception=True)` ?

Comment (sfc-gh-aalam) [reply tests/unit/test_server_connection.py:89]: added these

**Pull request merged**